pkg/dyninst/decode/types.go

```diff
@@ -0,0 +1,801 @@
+		if len(s.GroupType.Fields) < 2 {
+			return nil, errors.New("type map group has less than 2 fields: " + s.GroupType.Name)
+		}
+		slotsField := s.GroupType.Fields[1]
+		slotsFieldType, ok := d.program.Types[slotsField.Type.GetID()]
+		if !ok {
+			return nil, errors.New("type map slot field not found in types: " + s.GroupType.Name)
+		}
+		entryArray, ok := slotsFieldType.(*ir.ArrayType)
+		if !ok {
+			return nil, errors.New("type map slot field is not an array type: " + slotsFieldType.GetName())
+		}
+		noalgstructType, ok := entryArray.Element.(*ir.StructureType)
+		if !ok {
+			return nil, errors.New("type map entry array element is not a structure type: " + entryArray.Element.GetName())
+		}
+
+		if len(noalgstructType.Fields) < 2 {
+			return nil, errors.New("type map entry array element has less than 2 fields: " + entryArray.Element.GetName())
+		}
+		keyField := noalgstructType.Fields[0]
+		keyIrType, ok := d.program.Types[keyField.Type.GetID()]
+		if !ok {
+			return nil, fmt.Errorf("key type %s not found in types", keyField.Type.GetName())
+		}
+		keyType, err := d.getDecoderType(keyIrType)
+		if err != nil {
+			return nil, err
+		}
+
+		valueField := noalgstructType.Fields[1]
+		valueIrType, ok := d.program.Types[valueField.Type.GetID()]
+		if !ok {
+			return nil, fmt.Errorf("value type %s not found in types", valueField.Type.GetName())
+		}
+		valueType, err := d.getDecoderType(valueIrType)
+		if err != nil {
+			return nil, err
+		}
+		return &goSwissMapHeaderType{
+			GoSwissMapHeaderType: s,
+			keyType:              keyType,
+			valueType:            valueType,
+		}, nil
+
+	case *ir.BaseType:
+		return (*baseType)(s), nil
+	case *ir.StructureType:
+		return (*structureType)(s), nil
+	case *ir.ArrayType:
+		return (*arrayType)(s), nil
+	case *ir.VoidPointerType:
+		return (*voidPointerType)(s), nil
+	case *ir.GoSliceDataType:
+		return (*goSliceDataType)(s), nil
+	case *ir.GoStringHeaderType:
+		return (*goStringHeaderType)(s), nil
+	case *ir.GoStringDataType:
+		return (*goStringDataType)(s), nil
+	case *ir.GoMapType:
+		return (*goMapType)(s), nil
+	case *ir.GoHMapHeaderType:
+		return (*goHMapHeaderType)(s), nil
+	case *ir.GoHMapBucketType:
+		return (*goHMapBucketType)(s), nil
+	case *ir.GoSwissMapGroupsType:
+		return (*goSwissMapGroupsType)(s), nil
+	case *ir.GoChannelType:
+		return (*goChannelType)(s), nil
+	case *ir.GoEmptyInterfaceType:
+		return (*goEmptyInterfaceType)(s), nil
+	case *ir.GoInterfaceType:
+		return (*goInterfaceType)(s), nil
+	case *ir.GoSubroutineType:
+		return (*goSubroutineType)(s), nil
+	case *ir.EventRootType:
+		return (*eventRootType)(s), nil
+	default:
+		return nil, fmt.Errorf("unknown type %s", irType.GetName())
+	}
+}
+
+func (b *baseType) irType() ir.Type { return (*ir.BaseType)(b) }
+func (b *baseType) encodeValueFields(
+	decoder *Decoder,
+	enc *jsontext.Encoder,
+	dataItems map[typeAndAddr]output.DataItem,
+	currentlyEncoding map[typeAndAddr]struct{},
```

Comment:
It seems to me like all 4 of these fields are always the same for every call. I think we should do:

```
type decoderContext struct {
    d *Decoder
    enc *jsontext.Encoder
    dataItems map[typeAndAddr]output.DataItem
    currentlyEncoding map[typeAndAddr]struct{}
}
type decoderType interface {
	irType() ir.Type
	encodeValueFields(*decoderContext, data []byte) error
}

func encodeValue(
	ctx *decoderContext,
	decoderType decoderType,
	data []byte,
	valueType string,
) error {
```

We should be able to store this in the decoder, or, alternatively, just unify the two things